st3/lsp_utils/generic_client_handler.py

```diff
@@ -187,4 +188,25 @@ def on_ready(self, api: ApiWrapperInterface) -> None:
 
         :param api: The API instance for interacting with the server.
         """
-        pass
+        self._register_custom_server_event_handlers(api)
+
+    def _register_custom_server_event_handlers(self, api: ApiWrapperInterface) -> None:
+        """
+        Register decorator-style custom event handlers.
+
+        This method works as following steps:
+
+        1. Scan through all methods of this object.
+        2. If a method is decorated, it will has a "handler mark" attribute which is put by the decorator.
+        3. Register the method with wanted events, which are stored in the "handler mark" attribute.
+
+        :param api: The API instance for interacting with the server.
+        """
+        for _, func in inspect.getmembers(self, predicate=inspect.isroutine):
+            # client_event is like "notification", "request"
+            for client_event, handler_mark in HANDLER_MARKS.items():
+                event_registrator = getattr(api, "on_" + client_event, None)
+                if callable(event_registrator):
+                    server_events = getattr(func, handler_mark, [])  # type: List[str]
+                    for server_event in server_events:
+                        event_registrator(server_event, func)
```

Comment:
Does this allow for both decorators to be applied to the same function? That wouldn't make sense as the function signatures differ.

Reply:
yeah... at this moment I just assume those two are all `Callable[[Any, Any], None]`...

Reply:
I would really like the types to be correct and checked so that when decorating with `@notification_handler`, for example, would raise a type error if the function doesn't have a correct number of arguments (and maybe also correct types).

I'm not certain if it's possible for decorators to affect type checking though...

Reply:
> yeah... at this moment I just assume those two are all `Callable[[Any, Any], None]`...

Types being incorrect is one thing but I guess it would also crash at runtime since the number of arguments wouldn't match the passed number of arguments.

Reply:
Will investigate (my) tomorrow!

Reply:
If you enable pushes to this PR then I'll push a fix.

Reply:
![image](https://user-images.githubusercontent.com/6594915/99544565-6b5b4c00-29ef-11eb-8f0d-bb67fa5298f7.png)

Weird. I remember there used to be a "allow edit by maintainer" in the sidebar but there isn't.

Reply:
Maybe it's on the bottom, where the comment box is?

Reply:
Hmmm... can't find it

Reply:
Hey seriously I can't find it in https://docs.github.com/en/free-pro-team@latest/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork either now or while creating a PR...